src/executor/graph_executor.cc

```diff
@@ -676,38 +676,48 @@ void GraphExecutor::InitArguments(const nnvm::IndexedGraph& idx,
  * \brief If the requested ndarray's shape size is less than
  * the corresponding shared_data_array's shape size and the
- * storage type is default storage, reuse the memory allocation
+ * storage type is shareable, reuse the memory allocation
  * in shared_buffer; otherwise, create a zero ndarray.
+ * Shareable storages include both default storage and row_sparse storage
+ * if enable_row_sparse_sharing is `True`, otherwise default storage only.
  */
 NDArray ReshapeOrCreate(const std::string& name,
                         const TShape& dest_arg_shape,
                         const int dest_arg_dtype,
                         const NDArrayStorageType dest_arg_stype,
                         const Context& ctx,
-                        std::unordered_map<std::string, NDArray>* shared_buffer) {
-  if (dest_arg_dtype != kDefaultStorage) {
-    return InitZeros(dest_arg_stype, dest_arg_shape, ctx, dest_arg_dtype);
+                        std::unordered_map<std::string, NDArray>* shared_buffer,
+                        bool enable_row_sparse_sharing) {
+  bool stype_shareable = dest_arg_stype == kDefaultStorage;
+  if (enable_row_sparse_sharing) {
+    stype_shareable = stype_shareable || dest_arg_stype == kRowSparseStorage;
   }
   auto it = shared_buffer->find(name);
   if (it != shared_buffer->end()) {
-    if (it->second.shape().Size() >= dest_arg_shape.Size()) {  // memory can be reused
+    bool size_shareable = it->second.shape().Size() >= dest_arg_shape.Size();
+    if (size_shareable && stype_shareable) {  // memory can be reused
       CHECK_EQ(it->second.dtype(), dest_arg_dtype)
-        << "Requested arg array's dtype does not match the reusable ndarray";
-      CHECK_EQ(it->second.storage_type(), kDefaultStorage)
-               << "shared_buffer should only contain NDArrays with default storage type.";
+        << "Requested arg array's dtype does not match that of the reusable ndarray";
+      CHECK_EQ(it->second.storage_type(), dest_arg_stype)
+        << "Requested arg array's stype does not match that of the reusable ndarray";
       return it->second.Reshape(dest_arg_shape);
-    } else {
+    } else if (stype_shareable) {
       LOG(WARNING) << "Bucketing: data " << name << " has a shape " << dest_arg_shape
                    << ", which is larger than already allocated shape " << it->second.shape()
                    << ". Need to re-allocate. Consider putting default bucket key to be "
                    << "the bucket taking the largest input for better memory sharing.";
-      // the NDArrays in shared_buffer are guaranteed to be of default storage
+      // the NDArrays in shared_buffer are guaranteed to be of shareable storages
       it->second = InitZeros(dest_arg_stype, dest_arg_shape, ctx, dest_arg_dtype);
       return it->second;
+    } else {
+      // not shareable storage
+      return InitZeros(dest_arg_stype, dest_arg_shape, ctx, dest_arg_dtype);
     }  // arg_array.shape().Size() >= arg_shape.Size()
```

Comment:
line 704 is still performing sharing by updating `it->second `, it's just that its size is too small, so creating a bigger ndarray for sharing. Maybe the name `size_shareable ` is misleading?

Reply:
Yes, the comment "// arg_array.shape().Size() >= arg_shape.Size()" probably makes sense after if instead of else.   I don't think the name `size_shareable` is misleading.

Reply:
Oh I see you're referring to this comment. I'll move it